mcstatus/protocol/connection.py

```diff
@@ -221,119 +504,138 @@ def read(self, length: int) -> bytearray:
         self.socket.send(data)
 
-    def __del__(self):
-        try:
-            self.socket.close()
-        except Exception:  # TODO: Check what this actually excepts
-            pass
 
+class UDPSocketConnection(SocketConnection):
+    "UDP Connection to address. Default timeout is 3 seconds."
+    __slots__ = ("addr",)
 
-class UDPSocketConnection(Connection):
-    def __init__(self, addr: Address, timeout: float = 3):
-        Connection.__init__(self)
+    def __init__(self, addr: tuple[str, int], timeout: float = 3.0):
+        """Set self.addr to address, set self.socket to new socket,
+        AF_INET if IPv4, AF_INET6 otherwise."""
+        super().__init__()
         self.addr = addr
         self.socket = socket.socket(
             socket.AF_INET if ip_type(addr[0]) == 4 else socket.AF_INET6,
             socket.SOCK_DGRAM,
         )
         self.socket.settimeout(timeout)
 
-    def flush(self) -> bytearray:
-        raise NotImplementedError("UDPSocketConnection does not support flush()")
-
-    def receive(self, data: Union[BytesConvertable, SupportsBytes]) -> None:
-        raise NotImplementedError("UDPSocketConnection does not support receive()")
-
     def remaining(self) -> int:
+        "Return 65535."
         return 65535
 
     def read(self, length: int) -> bytearray:
+        "Return up to self.remaining() bytes. Length does nothing."
         result = bytearray()
         while len(result) == 0:
             result.extend(self.socket.recvfrom(self.remaining())[0])
         return result
 
-    def write(self, data: Union[Connection, bytes, bytearray]) -> None:
+    def write(self, data: Union[Connection, str, bytes, bytearray]) -> None:
+        "Use self.socket to send data to self.addr."
         if isinstance(data, Connection):
             data = bytearray(data.flush())
+        elif isinstance(data, str):
+            data = bytearray(data, "utf-8")
         self.socket.sendto(data, self.addr)
 
-    def __del__(self):
-        try:
-            self.socket.close()
-        except Exception:  # TODO: Check what this actually excepts
-            pass
 
+class TCPAsyncSocketConnection(BaseAsyncReadSyncWriteConnection):
+    "Asynchronous TCP connection to addr. Default timeout is 3 seconds."
+    __slots__ = ("reader", "writer", "timeout")
 
-class TCPAsyncSocketConnection(AsyncReadConnection):
-    # These will only be None until connect is called, ignore the None type assignment
-    reader: asyncio.StreamReader = None  # type: ignore[assignment]
-    writer: asyncio.StreamWriter = None  # type: ignore[assignment]
-    timeout: float = None  # type: ignore[assignment]
+    def __init__(self) -> None:
+        # These will only be None until connect is called, ignore the None type assignment
+        self.reader: asyncio.StreamReader = None  # type: ignore[assignment]
+        self.writer: asyncio.StreamWriter = None  # type: ignore[assignment]
+        self.timeout: float = 3.0
```

Comment:
None would mean needing to ignore a type assignment though.

Reply:
Yeah, but that's not a huge issue considering we already type-ignore the reader and writer variables, there's just no point in assigning some value here if it's always expected to be overridden later.

Reply:
Changed to 0 in c92282f, is that sufficient?

Reply:
I'd prefer `None`, since specifying a number here is misleading, suggesting that the timeout is hard-coded to certain value, whereas just setting it to None and type-ignoring with that comment above explaining why these are ignored is a bit better way to handle this.

Reply:
Changed in c09d857